src/v/cluster/tx_gateway_frontend.cc

```diff
@@ -581,8 +595,36 @@ ss::future<cluster::init_tm_tx_reply> tx_gateway_frontend::init_tm_tx_locally(
       [tx_id, transaction_timeout_ms, timeout](tx_gateway_frontend& self) {
           return ss::with_gate(
             self._gate, [tx_id, transaction_timeout_ms, timeout, &self] {
-                return self.do_init_tm_tx(
-                  tx_id, transaction_timeout_ms, timeout);
+                return self.with_stm(
+                  [tx_id, transaction_timeout_ms, timeout, &self](
+                    checked<ss::shared_ptr<tm_stm>, tx_errc> r) {
+                      if (!r) {
+                          return ss::make_ready_future<
+                            cluster::init_tm_tx_reply>(
+                            cluster::init_tm_tx_reply{r.error()});
+                      }
+                      auto stm = r.value();
+                      return stm->read_lock().then(
+                        [&self, stm, tx_id, transaction_timeout_ms, timeout](
+                          ss::basic_rwlock<>::holder unit) {
+                            return with(
+                                     stm,
+                                     tx_id,
+                                     "init_tm_tx",
+                                     [&self,
+                                      stm,
+                                      tx_id,
+                                      transaction_timeout_ms,
+                                      timeout]() {
+                                         return self.do_init_tm_tx(
+                                           stm,
+                                           tx_id,
+                                           transaction_timeout_ms,
+                                           timeout);
+                                     })
+                              .finally([u = std::move(unit)] {});
```

Comment:
do you also need to keep the `stm` object alive here? presumably `unit` holds a raw pointer to a semaphore in the stm?

Reply:
Maybe it is time to rewrite it to coroutines style?

Reply:
the object behind the stm shared pointer is kept alive inside with_stm

Reply:
inside `with_stm` i don't see any lifetime extension. am i missing that?